Cython/Compiler/Dataclass.py

```diff
@@ -0,0 +1,697 @@
+            continue
+        entry = node.scope.lookup(name)
+        if entry.annotation:
+            annotation = u": %s" % entry.annotation.string.value
+        else:
+            annotation = u""
+        assignment = u''
+        if field.default is not MISSING or field.default_factory is not MISSING:
+            seen_default = True
+            if field.default_factory is not MISSING:
+                ph_name = default_factory_placeholder
+            else:
+                ph_name = get_placeholder_name()
+                placeholders[ph_name] = field.default  # should be a node
+            assignment = u" = %s" % ph_name
+        elif seen_default:
+            error(entry.pos, ("non-default argument '%s' follows default argument "
+                              "in dataclass __init__") % name)
+            return "", {}, []
+
+        args.append(u"%s%s%s" % (name, annotation, assignment))
+
+        if field.is_initvar:
+            continue
+        elif field.default_factory is MISSING:
+            if field.init.value:
+                function_body_code_lines.append(u"    %s.%s = %s" % (selfname, name, name))
+        else:
+            ph_name = get_placeholder_name()
+            placeholders[ph_name] = field.default_factory
+            if field.init.value:
+                # close to:
+                # def __init__(self, name=_PLACEHOLDER_VALUE):
+                #     self.name = name_default_factory() if name is _PLACEHOLDER_VALUE else name
+                function_body_code_lines.append(u"    %s.%s = %s() if %s is %s else %s" % (
+                    selfname, name, ph_name, name, default_factory_placeholder, name))
+            else:
+                # still need to use the default factory to initialize
+                function_body_code_lines.append(u"    %s.%s = %s()"
+                                  % (selfname, name, ph_name))
+
+    args = u", ".join(args)
+    func_def = u"def __init__(%s):" % args
+
+    code_lines = [func_def] + (function_body_code_lines or ["pass"])
+
+    if node.scope.lookup("__post_init__"):
+        post_init_vars = ", ".join(name for name, field in fields.items()
+                                   if field.is_initvar)
+        code_lines.append("    %s.__post_init__(%s)" % (selfname, post_init_vars))
+    return u"\n".join(code_lines), placeholders, []
+
+
+def generate_repr_code(repr, node, fields):
+    """
+    The CPython implementation is just:
+    ['return self.__class__.__qualname__ + f"(' +
+                     ', '.join([f"{f.name}={{self.{f.name}!r}}"
+                                for f in fields]) +
+                     ')"'],
+
+    The only notable difference here is self.__class__.__qualname__ -> type(self).__name__
+    which is because Cython currently supports Python 2.
+    """
+    if not repr or node.scope.lookup("__repr__"):
+        return "", {}, []
+    code_lines = ["def __repr__(self):"]
+    strs = [u"%s={self.%s}" % (name, name)
```

Comment:
I think you shold use %r instead of %s here:

```suggestion
    strs = [u"%s={self.%r}" % (name, name)
```

The original CPython implementation does `{{self.{f.name}!r}}` (notice the `!r` for the repr form, not the plain str).

E.g. I have a a class like this

```
@dataclass
@cython.cclass
class Point:
    x: float
    y: float
    name: str
```

When I look at the cython-generated repr() I get the string attribute values without the quotes

```
Point(x=1.0, y=2.0, name=abc)
```

which means I can't eval the repr (which I think it's the point of using `repr()` or `!r` instead of just str().

Reply:
Thanks - should be fixed now.